openprescribing/pipeline/management/commands/fetch_and_import_ncso_concessions.py

```diff
@@ -223,4 +266,122 @@ def regularise_ncso_name(name):
     # Remove spaces around slashes
     name = re.sub(" */ *", "/", name)
 
+    # dm+d names often have the units appended after the pack size (e.g. "12 tablets"),
+    # we want to remove the units leaving just the "12"
+    name = re.sub(r"(\d+) [^\d]+$", r"\1", name)
+
     return name
+
+
+def get_vmpp_id_from_previous_concession(drug, pack_size):
+    previous_vmpp_ids = list(
+        NCSOConcession.objects.filter(
+            drug=drug, pack_size=pack_size, vmpp__isnull=False
+        ).values_list("vmpp_id", flat=True)
+    )
+    if previous_vmpp_ids:
+        return get_single_item(previous_vmpp_ids)
+
+
+def insert_or_update(items):
+    # Sort by published date so most recent prices are always applied last
+    items = sorted(items, key=lambda i: i["publish_date"])
+
+    new_concession_count = 0
+
+    with transaction.atomic():
+        for item in items:
+            if item["vmpp_id"] is not None:
+                _, created = NCSOConcession.objects.update_or_create(
+                    date=item["date"],
+                    vmpp_id=item["vmpp_id"],
+                    defaults=dict(
+                        drug=item["drug"],
+                        pack_size=item["pack_size"],
+                        price_pence=item["price_pence"],
+                    ),
+                )
+            else:
+                # Unmatched concessions get inserted as placeholder entries, ready for
+                # manual reconciliation
+                _, created = NCSOConcession.objects.update_or_create(
+                    date=item["date"],
+                    drug=item["drug"],
+                    pack_size=item["pack_size"],
+                    vmpp_id=None,
+                    defaults=dict(
+                        price_pence=item["price_pence"],
+                    ),
+                )
+
+            if created:
```

Comment:
Is it helpful to log whether this was a matched or unmatched new concession?

Reply:
I think it's OK as is. We report the unmatched concession elsewhere.